limiter.go

```diff
@@ -0,0 +1,148 @@
+//
+// Supported options are:
+//   - WithNumberBuckets: Sets the number of buckets used for expiring quotas.
+//     This must be greater than zero, and defaults to DefaultNumberBuckets. A
+//     larger number of buckets can increase the efficiency at which expired
+//     quotas are deleted to free up space. However, it does also marginally
+//     increase the amount of memory needed, and can increase the frequency
+//     in which the delete routine runs and must acquire a lock.
+func NewLimiter(limits []*Limit, maxSize int, o ...Option) (*Limiter, error) {
+	const op = "rate.NewLimiter"
+
+	switch {
+	case len(limits) <= 0:
+		return nil, fmt.Errorf("%s: %w", op, ErrEmptyLimits)
+	}
+
+	byKey := make(map[string]*Limit, len(limits))
+
+	var maxEntryTTL time.Duration
+	for _, l := range limits {
+		if !l.IsValid() {
+			return nil, fmt.Errorf("%s: %w", op, ErrInvalidLimit)
+		}
+		key := getKey(l.Resource, l.Action, string(l.Per))
+		if _, ok := byKey[key]; ok {
+			return nil, fmt.Errorf("%s: %s %s %s: %w", op, l.Resource, l.Action, l.Per, ErrDuplicateLimit)
+		}
+		byKey[key] = l
+		if l.Period > maxEntryTTL {
+			maxEntryTTL = l.Period
+		}
+	}
+
+	// TODO: handle special case where all of the provided limits have Unlimited = true.
+	// If this is the case, we can skip the creation of a quotaFetcher
+
+	s, err := newExpireableStore(maxSize, maxEntryTTL, o...)
+	if err != nil {
+		return nil, fmt.Errorf("%s: %w", op, err)
+	}
+
+	l := &Limiter{
+		limits:       byKey,
+		quotaFetcher: s,
+	}
+
+	return l, nil
+}
+
+// Allow checks if a request for the given resource and action should be allowed.
+// A request is not allowed if:
+//   - Any of the associated quotas have been exhausted.
+//   - A new quota needs to be stored but there is no available space to store it.
+//   - There is no corresponding limit for the resource and action.
+func (l *Limiter) Allow(resource, action string) (allowed bool, quota *Quota, err error) {
+	l.mu.RLock()
+	defer l.mu.RUnlock()
+
+	var ok bool
+	var limit *Limit
+	var q *Quota
+	var key string
+	allowed = true
+
+	allowOrder := []LimitPer{LimitPerTotal}
+
+	quotas := make(map[LimitPer]*Quota, len(allowOrder))
+	keys := map[LimitPer]string{
+		LimitPerTotal: string(LimitPerTotal),
+	}
+
+	for per, id := range keys {
+		key = getKey(resource, action, string(per))
+		limit, ok = l.limits[key]
+		if !ok {
+			allowed = false
+			err = ErrLimitNotFound
+			return
+		}
+
+		q, err = l.quotaFetcher.fetch(id, limit)
+		if err != nil {
+			allowed = false
+			return
+		}
+
+		if q.Remaining() <= 0 {
+			allowed = false
+			quota = q
+			return
+		}
+
+		quotas[LimitPerTotal] = q
+	}
+
+	for _, q := range quotas {
+		q.Consume()
+		if quota == nil || q.Remaining() < quota.Remaining() {
+			quota = q
+		}
+	}
+
+	return
+}
+
+// Shutdown stops a Limiter. After calling this, any future calls to Allow
+// will result in ErrStopped being returned.
+func (l *Limiter) Shutdown() error {
+	return l.quotaFetcher.shutdown()
```

Comment:
Can this `Shutdown()` method ever return an error since the `expirableStore` `shutdown()` method never does?

Reply:
Currently it would not return an error, but other implementations of `quotaFetcher` might. So it seems prudent to have this return an error instead of needing to change the exported method later.